### PR TITLE
Update AI agents instructions

### PR DESCRIPTION
**Pull request by advieser**
Based on https://github.com/mlr-org/mlr3infrastructure, we are updating our own AI agents instructions.

The current challenge is that the different agent models, follow different conventions for extra instructions.
* Codex reads `AGENTS.md`
  - Maximum of 32 KiB by default for all discovered project instructions.
  - AFAICT, there's no way to force `extra-rules` into context. We could only a) put all the information into `AGENTS.md`or b) instruct the agent in the `AGENTs.md` to read those files.
* Claude Code reads `CLAUDE.md`. 
  - Right now this is a symlink, but we could also use CLAUDE's import (`@`) functionality, if we want to add Claude-specific instructions as well.
  - Maximum of 4 MiB. Shorter files produce better adherence.
  - Additional information should be in `.claude/rules`. We could have project-level directory `extra-rules` that gets symlinked here. 
 
This is WIP. `AGENTS.md` is the current master file. Lines that already get covered in the WIP setup (either in `AGENTS_new.md` - to be renamed later - or in `extra_rules/`) got deleted to keep track easier. 
Before merging this PR, make sure that the `CLAUDE.md` symlink links to the correct file.

Open question: Why does `extra-rules/mlr3.md` repeat information from `AGENTS.md`?

### CODE REVIEW

File: extra-rules/mlr3.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+- `Measure` > `MeasureClassif` / `MeasureRegr` / `MeasureSimilarity`
+- `Resampling` > `ResamplingCV`, `ResamplingHoldout`, etc.
+- `DataBackend` > `DataBackendDataTable`, `DataBackendCbind`, etc.
+- `Prediction` > `PredictionClassif` / `PredictionRegr`
+
+#### File naming
+
+- One R6 class per file, named exactly as the class: `LearnerClassifRpart.R` contains `LearnerClassifRpart`.
+- Named dataset tasks use an underscore: `TaskClassif_iris.R`.
+- Dictionary files: `mlr_learners.R`, `mlr_tasks.R`, etc.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: extra-rules/mlr3.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+| `mlr_measures`        | `msr()` / `msrs()`   | `msr("classif.ce")`              |
+| `mlr_resamplings`     | `rsmp()` / `rsmps()` | `rsmp("cv", folds = 5)`          |
+| `mlr_task_generators` | `tgen()` / `tgens()` | `tgen("friedman1")`              |
+
+Every new object **must** be registered at the bottom of its file:
+
+```r
+#' @include mlr_learners.R
+mlr_learners$add("classif.rpart", function() LearnerClassifRpart$new())
+```
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
| `mlr_learners`        | `lrn()` / `lrns()`   | `lrn("classif.rpart", cp = 0.1)` |
| `mlr_pipeops`           | `po()` / `pos()`   | `po("pca")`                    |

etc. New objects must be registered in the respective dictionary; this works differently for external (e.g. `mlr_learners`) than for package-owned (`mlr_pipeops`) dictionaries.
```

File: extra-rules/mlr3.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+#### Hyperparameters (paradox)
+
+Parameters are defined with `paradox::ps()` and must be tagged `"train"` or `"predict"`:
+
+```r
+ps = ps(
+  cp = p_dbl(0, 1, default = 0.01, tags = "train"),
+  keep_model = p_lgl(default = FALSE, tags = "train")
+)
+```
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
Parameters are defined with `paradox::ps()` and should usually be tagged `"train"`, `"predict"`, or both.
```

File: extra-rules/mlr3.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+In `.train()` / `.predict()`, retrieve values with `self$param_set$get_values(tags = "train")`.
```
Comment (mb706) [inline extra-rules/mlr3.md:22]:
```suggestion
In `.train()` / `.predict()`, retrieve values with `self$param_set$get_values(tags = "train")`.
Use additional tags for additional grouping (e.g. parameters that get passed to different library functions.
```

File: AGENTS.md
Comment (advieser) [inline AGENTS.md]:
```suggestion

```
Covered by commit from mb706.

Comment (advieser) [inline AGENTS.md]:
```suggestion

```
Covered by commit by mb706

File: extra-rules/mlr3.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+There is a distinction between `default` and `init` values:
+- `default` describes the behavior when a parameter is not set at all (i.e., the upstream function's default). It is informational only.
+- `init` (via `p_xxx(init = ...)`) sets the parameter to a value upon construction. Use this when the mlr3 default should differ from the upstream default.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
- `init` (via `p_xxx(init = ...)`) sets the parameter to a value upon construction. Use this when our own default should differ from the upstream default, or when upstream is `"required"` but there is a reasonable default. For functions that we implement in our own package, this is typically the way we go. The `init()` functionality is new, some old code does the initialization differently, via `ps$values = ...`; do not copy the old style.
```

File: extra-rules/mlr3.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+- A parameter tagged `"required"` causes an error if not set. A required parameter cannot have a `default` (that would be contradictory).
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
- A parameter tagged `"required"` causes an error if not set. A required parameter cannot have a `default` (that would be contradictory). For functionality that we implement ourselves, we usually use the `"required"` route.
```

File: AGENTS.md
Comment (advieser) [inline AGENTS.md]:
```suggestion

```
Covered by `extra-rules/mlr3.md`.

File: extra-rules/mlr3.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+For read-only fields, call `assert_ro_binding(rhs)` to raise an error on any assignment attempt:
+
+```r
+task_type = function(rhs) {
+  assert_ro_binding(rhs)
+  private$.data$task_type
+},
+```
+
+The private slot must be declared in the `private` list as `.id = NULL` (or an appropriate default).
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: extra-rules/mlr3.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+For read-only fields, call `assert_ro_binding(rhs)` to raise an error on any assignment attempt:
+
+```r
+task_type = function(rhs) {
+  assert_ro_binding(rhs)
+  private$.data$task_type
+},
+```
```
Comment (mb706) [inline extra-rules/mlr3.md]:
````suggestion
Public fields on R6 classes are exposed as active bindings backed by a private `.field`.

For mutable fields, the binding returns the private value when called without arguments and validates the new value with an `assert_*()` call when set. For read-only fields, call `assert_ro_binding(rhs)` to raise an error on any assignment attempt:

```r
id = function(rhs) {
  if (missing(rhs)) {
    return(private$.id)
  }
  private$.id = assert_id(rhs)
},
task_type = function(rhs) {
  assert_ro_binding(rhs)
  private$.data$task_type
},
```
````

File: extra-rules/mlr3.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+The private slot must be declared in the `private` list as `.id = NULL` (or an appropriate default).
+
+#### Core dependencies
+
+`data.table`, `checkmate`, `mlr3misc`, `paradox`, `R6`, and `cli` are imported wholesale. Use their functions directly without `::`. Key mlr3misc utilities: `map()`, `map_chr()`, `invoke()`, `calculate_hash()`, `str_collapse()`, `%nin%`, `%??%`.
+
+#### Error handling
+
+Use structured error/warning functions from mlr3misc: `error_config()`, `error_input()`, `error_learner_train()`, `error_learner_predict()`, `warning_config()`, `warning_input()`. These support `sprintf`-style formatting.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: AGENTS.md
Comment (advieser) [inline AGENTS.md]:
```suggestion

```
covered in mlr3.md

File: extra-rules/mlr3.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+#### Reflections
+
+`mlr_reflections` is an environment that stores allowed types, properties, and roles. Extension packages modify it to register new task types. Check it when adding new properties or feature types.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: AGENTS.md
Comment (advieser) [inline AGENTS.md]:
```suggestion

```
covered in mlr3.md

Comment (advieser) [inline AGENTS.md]:
```suggestion

```
covered in mlr3.md

Comment (advieser) [inline AGENTS.md]:
```suggestion

```

File: extra-rules/mlr3.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+- Tests for `R/{name}.R` go in `tests/testthat/test_{name}.R`.
+- All new code should have an accompanying test.
+- If there are existing tests, place new tests next to similar existing tests.
+- Strive to keep your tests minimal with few comments.
+- The full test suite takes a long time. Only run tests relevant to your changes with `devtools::test(filter = '^{name}')`.
+- New learners must pass `run_autotest()` and `run_paramtest()`.
+- Use shared assertion helpers: `expect_learner()`, `expect_task()`, `expect_resampling()`, `expect_measure()`, `expect_prediction()`.
+- Shared test infrastructure lives in `inst/testthat/` and is sourced by extension packages too.
+- Use `skip_if_not_installed(<package_name>)` to skip tests that require suggested packages.
```
Comment (mb706) [inline extra-rules/mlr3.md:48]:
```suggestion

- The full test suite takes a long time. Only run tests relevant to your changes with `devtools::test(filter = '^{name}')`.
- New PipeOps must pass `expect_pipeop_class`, or, in case of preprocessing PipeOps that inherit from `PipeOpTaskPreproc[Simple]`,  `expect_datapreproc_pipeop_class`. The latter calls the former, and both call `expect_pipeop`, which should not be necessary by itself in most cases.
- Use shared assertion helpers: `expect_learner()`, `expect_task()`, `expect_resampling()`, `expect_measure()`, `expect_prediction()`.
- Shared test infrastructure lives in `inst/testthat/` and is sourced by extension packages too.
- Use `skip_if_not_installed(<package_name>)` to skip tests that require suggested packages.
```

File: AGENTS.md
Comment (advieser) [inline AGENTS.md]:
```suggestion

```
covered in mlr3.md

File: extra-rules/mlr3.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+### Documentation
+
+- Every user-facing function should be exported and have roxygen2 documentation.
+- Wrap roxygen comments at 120 characters.
+- Write one sentence per line.
+- If a sentence exceeds the limit, break at a comma, "and", "or", "but", or other appropriate point.
+- Internal functions should not have roxygen documentation.
+- Whenever you add a new (non-internal) documentation topic, also add the topic to `_pkgdown.yml`.
+- Always re-document the package after changing a roxygen2 comment.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
- Always rerefresh documentation via `roxygen2::roxygenize()` after changing a roxygen2 comment.
```

File: AGENTS.md
Comment (advieser) [inline AGENTS.md]:
```suggestion
In rare cases, this can differ from default, e.g. if the underlying default behaviour is suboptimal for the use for preprocessing (e.g. it stores training data unnecessarily by default).
```
covered in mlr3.md

File: extra-rules/mlr3.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+- Use `pkgdown::check_pkgdown()` to check that all topics are included in the reference index.
+- Don’t hand-edit generated artifacts: `man/`, or `NAMESPACE`.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: extra-rules/mlr3.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+- Roxygen templates live in `man-roxygen/` (e.g., `@template learner`, `@template param_id`). Use `@templateVar` to pass values.
+- Bibliographic references go in `R/bibentries.R` and are cited with `` `r format_bib("key")` ``.
+- Man page names for dictionary objects follow `mlr_learners_classif.rpart`, `mlr_tasks_iris`, etc.
+- When you write examples, make sure they work.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: extra-rules/mlr3.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+- Wrap examples that use suggested packages in `if (mlr3misc::require_namespaces(<package_name>, quietly = TRUE)) {..}` blocks.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
- Wrap parts of examples that use suggested packages in `if (mlr3misc::require_namespaces(<package_names_vector>, quietly = TRUE)) {..}` blocks. When essentially the entire example needs the package, use `#' @examplesIf mlr3misc::require_namespaces(<package_names_vector>, quietly = TRUE)` instead of `#' @examples`, which inserts the `if` automatically.
```

File: extra-rules/mlr3.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+- Every user-facing change should be given a bullet in `NEWS.md`. Do not add bullets for small documentation changes or internal refactorings.
+- Each bullet should briefly describe the change to the end user and mention the related issue in parentheses.
+- A bullet can consist of multiple sentences but should not contain any new lines (i.e. DO NOT line wrap).
+- If the change is related to a function, put the name of the function early in the bullet.
+- Order bullets alphabetically by function name. Put all bullets that don't mention function names at the beginning.
+
+### GitHub
+
+- If you use `gh` to retrieve information about an issue, always use `--comments` to read all the comments.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: extra-rules/mlr3.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+### Writing
+
+- Use sentence case for headings.
+- Use US English.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: extra-rules/mlr3.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+Fix spelling, grammar, and other minor problems without asking the user. Label any unclear, confusing, or ambiguous sentences with a FIXME comment.
+
+Only report what you have changed.
+
+### References
+
+- [mlr3book](https://mlr3book.mlr-org.com/) — comprehensive guide to the mlr3 ecosystem.
+- [mlr3misc](https://github.com/mlr-org/mlr3misc) — helper functions used throughout the codebase.
+- [paradox](https://github.com/mlr-org/paradox) — hyperparameter/configuration space definitions.
```
Comment (mb706) [inline extra-rules/mlr3.md]:
```suggestion
```

File: AGENTS_new.md [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+## Key commands
+
+```
+# To run code
+Rscript -e "devtools::load_all(); code"
+
+# To run all tests
+Rscript -e "devtools::test()"
+
+# To run all tests for files starting with {name}
```
Comment (mb706) [inline AGENTS_new.md]:
```suggestion
# To run all tests for files starting with {name}; omit filter to run all tests (expensive).
```

File: AGENTS_new.md [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+* Use these mlr3misc utilities when appropriate:
+  `map()`, `map_chr()`, `invoke()`, `calculate_hash()`, `str_collapse()`, `%nin%`, `%??%`.
+* Before implementing something, read similar existing files first to match the established patterns.
+* Always use `# nolint next` to disable linters for the next line instead of `# nolint` on the same line.
+
+## File structure and naming
+
+* Name the file as the most important contained function / class
+* No whitespaces, no special chart in filenames
+* Usually one large function / class, per file, but adding multiple smaller helpers is ok
```
Comment (mb706) [inline AGENTS_new.md]:
```suggestion
* Name the file as the most important contained function / class
* Usually one large function / class, per file; if auxiliary functions pertain almost exclusively to that, it they should go in the same file, not an auxiliary file.
```

File: AGENTS_new.md [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+## Collation order
+
+* Derived classes must declare `#' @include ParentClass.R` in their roxygen header.
+  This controls the `Collate:` field in DESCRIPTION so base classes load before derived classes.
+
+## Core dependencies
+* Use `checkmate` for arg-checks
+* Use `data.table` for efficient table structures
+* For OOP-style use `R6`
+* Use `cli` to format messages, warnings, errors and prints
```
Comment (mb706) [inline AGENTS_new.md:60]:
```suggestion
## Core dependencies

* Use `checkmate` for arg-checks
* Use `data.table` instead of `data.frame` or tibbles.
* Use `cli` to format messages, warnings, errors and prints
```

File: AGENTS_new.md [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+
+## Testing
+
+* Tests for `R/{name}.R` go in `tests/testthat/test_{name}.R`.
+* All new code should have an accompanying test.
+* If there are existing tests, place new tests next to similar existing tests.
+* Strive to keep your tests minimal with few comments.
+* The full test suite takes a long time. Only run tests relevant to your changes with `devtools::test(filter = '^{name}')`.
+* Read the additional important helpers in `inst/testthat/helper_functions.R` to understand our `PipeOpTaskPreproc` auto-test framework.
```
Comment (mb706) [inline AGENTS_new.md:69]: todo: merge with testing part above

File: AGENTS_new.md [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+- If you use `gh` to retrieve information about an issue, always use `--comments` to read all the comments.
+
+## Natural Language
+
+- The following applies to all natural language text, so docs, comments, NEWS, etc, but not code
+- Use American English
+- Use the Oxford comma
+- Do not capitalize normal nouns or method names. "Bayesian" is capitalized, "random forest" is not.
+- Use cspell to check against typos, and add needed words to .cspell/project-words.txt if reasonable
+
```
Comment (mb706) [inline AGENTS_new.md]:
```suggestion
```
